serve/tests/unittest/test_engine_with_samplers.py

```diff
@@ -362,47 +259,65 @@ def _test_logprobs(
             max_tokens=300,
             stop=None,
             ignore_eos=True,
-            top_logprobs=top_logprobs,
-            logprobs=True
-        ) for n in range(num_requests)
+            top_logprobs=random.randint(1, 5),
+            logprobs=True,
+        )
+        for n in range(num_requests)
     ]
     engine.add(requests)
 
     generated = ["" for _ in range(num_requests)]
-
     while engine.has_pending_requests():
         results = engine.step()
         for res in results.outputs:
             assert len(res.sequences) == 1
             seq = res.sequences[0]
-
-            assert seq.finish_reason is not None or len(seq.logprob_info[0].top_logprobs) == top_logprobs
+            assert (
+                seq.finish_reason is not None
+                or len(seq.logprob_info[0].top_logprobs)
+                == requests[int(res.request_id)].sampling_params.top_logprobs
+            )
 
             if seq.is_finished:
-                assert seq.num_generated_tokens == requests[int(res.request_id)].stopping_criteria.max_tokens
+                assert (
+                    seq.num_generated_tokens
+                    == requests[int(res.request_id)].stopping_criteria.max_tokens
+                )
                 assert seq.finish_reason == FinishReason.Length
             else:
                 generated[int(res.request_id)] += seq.delta
+        break
 
-    if use_staging_engine:
-        engine.stop()
 
 if __name__ == "__main__":
     parser = get_default_mlc_serve_argparser("test engine with samplers")
     args = parser.parse_args()
     postproc_mlc_serve_args(args)
+    max_num_batched_tokens = 2048
 
-    _test_max_tokens(args.model_artifact_path, use_staging_engine=True)
-    _test_max_tokens(args.model_artifact_path, use_staging_engine=False)
-    _test_ignore_eos(args.model_artifact_path, use_staging_engine=True)
-    _test_ignore_eos(args.model_artifact_path, use_staging_engine=False)
-    _test_stop(args.model_artifact_path, use_staging_engine=False)
-    _test_stop(args.model_artifact_path, use_staging_engine=True)
-    _test_logprobs(args.model_artifact_path, use_staging_engine=True)
-    _test_logprobs(args.model_artifact_path, use_staging_engine=False)
+    # Test staging engines
+    staging_engine = create_engine(
+        args.model_artifact_path, max_num_batched_tokens, use_staging_engine=True
+    )
+    _test_max_tokens(staging_engine)
+    _test_ignore_eos(staging_engine)
+    # TODO (@sunggg): There is something stateful.
```

Comment:
Interestingly, `_test_logrpobs` does not work when `_test_stop` is enabled, seems like `_test_stop` has some statefulness that affects the next tests. I see the following test starts after generating one extra token at the beginning which I don't understand where it comes from atm. @yelite, can you take a look?

Reply:
May I ask if the order of the unit tests impact those results? Like if we move `test_stop` after `test_logprobs` would that work?

Reply:
Yeah, if I change the order, it works. So I suspect there is something going on with `stop`.

Reply:
It turns out that it's not really a bug in the InferenceEngine, although such out-of-sync is undesirable from the perspective of implementation. What really causes trouble is that different tests send requests with same request ids. Here `_test_stop` sends request id 2, and `_test_logprobs` also has a request id 2. A quick fix for this test failure is to make sure request ids are unique. I will fix the underlying out-of-sync in the engine refactoring.